[1.21.10] Configuration API Overhaul

Dragon-Seeker opened the pull request:
Primarly this PR attempts to resolve various pain points found or head of when working with owo-config:

- [x] Better handling for various Data types that are commonly used (Supported Types below)
- [x] Adjust current builder functions that other users of the lib can use to remove the need for reflection on field directly
- [x] Add ability for different constraints for apply and input
- [x] Adjust sliders for changes that chyz desired 
- [x] Config Sync
	- [x] Inform server of changes that any client makes when connected to the server
	- [x] Inform server of current connecting clients config values
- [x] Server Config Management
	- [x] Ability to open server config if the given user has OP privilege's
	- [x] Allow for the ability to reload server config from command
- [x] Multiple Configs under one given Mod
- [x] Pull changes from Chyz dealing with Sliders to allow for more customizability
- [x] UI Rework
	- [x] Wrapping option labels to prevent them from being unreadable if too long
	- [x] Adjust the slide out section selection to be an overlay instead of part of the primary config layout
	- [x] Support Config switching similar to Create New World Screen top bar
- [x] Development Translation Key Utilities's
	- [x] Print out the entire configs worth of translations with the current existing translations filled in
	- [x] Ability to dump individual missing translations
		- [ ] Make custom font for better looking clipboard

### Supported Types:
- [x] All Numbers (Integers and Floats)
- [x] String
- [x] Struct*
- [x] Record
- [x] Collections
	- [x] List
		- [x] String
		- [x] Identifier
		- [x] Numbers
		- [x] Struct
		- [x] Record
	- [x] Set
		- [x] String
		- [x] Identifier
		- [x] Numbers
		- [x] Struct
		- [x] Record
	- [x] Map
		- [x] Identifier - Identifier
		- [x] Identifier - String
		- [x] Identifier - Number
		- [x] Identifier - Enum
		- [x] String - Identifier
		- [x] String - String
		- [x] String - Number
		- [x] String - Enum
		- [x] Number - Identifier
		- [x] Number - String
		- [x] Number - Number
		- [x] Number - Enum

Dragon-Seeker committed:
Initial work to adjust and update owo config

- More supported types
- Better config Sync
- Server Config Management
- Ability to handle more than one config instance
- Slider Adjustments from Chyz
- Adjust how section headers are on the side
- Ability to switch between config instances
- Simple utility to copy translation key for missing translation in config
- Rework underling code to handle other types of objects and give better access to customize certain option calls without reflection

# Conflicts:
#	src/main/java/io/wispforest/owo/Owo.java
#	src/main/java/io/wispforest/owo/client/OwoClient.java
#	src/main/java/io/wispforest/owo/config/ConfigWrapper.java
#	src/main/java/io/wispforest/owo/config/ui/ConfigScreen.java
#	src/main/java/io/wispforest/owo/config/ui/ConfigScreenProviders.java
#	src/main/java/io/wispforest/owo/ui/base/BaseParentComponent.java
#	src/main/java/io/wispforest/owo/ui/component/LabelComponent.java
#	src/main/java/io/wispforest/owo/ui/container/Containers.java
#	src/main/java/io/wispforest/owo/ui/container/ScrollContainer.java
#	src/main/java/io/wispforest/owo/ui/core/OwoUIDrawContext.java
#	src/main/java/io/wispforest/owo/ui/core/Surface.java
#	src/main/java/io/wispforest/owo/ui/util/MatrixStackTransformer.java
#	src/main/resources/owo.accesswidener
#	src/testmod/java/io/wispforest/uwu/client/SelectUwuScreenScreen.java

Dragon-Seeker committed:
Update owo UI xml validation

Dragon-Seeker committed:
Add renderdoc test run

Dragon-Seeker committed:
Adjust Game rule disabling feature to prevent constant disabling instead only on initial creation

Dragon-Seeker committed:
Fix issues with passing Wrapper component instead of clickable widget

Dragon-Seeker committed:
Add some wisdom

Dragon-Seeker committed:
Fix issue with button appearing for configuration with owolib

Dragon-Seeker committed:
Change Config Constaints, add translation dumping, rework option annotation access

- Update Config Constraints to allow for different ones for input and apply
- Add proper way of dumping translation data using `ConfigTranslationHelper`
- Rework option annotation access around `AnnotatedElement`

Dragon-Seeker committed:
Adjust Client/Server switch button with a Icon and implement such as a `ToggleButton`

 Conflicts:
	src/main/java/io/wispforest/owo/ui/component/ButtonComponent.java
	src/main/java/io/wispforest/owo/ui/core/OwoUIDrawContext.java

Dragon-Seeker committed:
Add `ReloadRequired` annotation

Dragon-Seeker committed:
Update ConfigSynchronizer to use Identifier with use of OwoNetworking

Dragon-Seeker committed:
Adjust commands related to config access/reload

Dragon-Seeker committed:
Fix issue with not properly transferring parent screen from opening server config

Dragon-Seeker committed:
Fix issues with not checking if a given message can be sent within OwoNetChannel

Dragon-Seeker committed:
Properly send error for when OwoUIAdapater errors in `render` for BaseOwoScreen

Dragon-Seeker committed:
Update Test mod config wrappers with new field for reload and other stuff

Dragon-Seeker committed:
Fix issues with client mixins in common side leading to annoying errors

Dragon-Seeker committed:
Use `ConfigIdentifierArgumentType` arg for uwu debug command

Dragon-Seeker committed:
Add uwu mod icon

Dragon-Seeker committed:
Add ability to dump translations to log from debug button instead of after screen build

Dragon-Seeker committed:
Update struct option container to clean up `addOptionComponent` method

Dragon-Seeker committed:
Add `-XX:+AllowEnhancedClassRedefinition`

Dragon-Seeker committed:
Fix issues with porting to 1.21.6

Dragon-Seeker committed:
Remove `configName` and unused classes

Dragon-Seeker committed:
Adjust options method names for label and tooltip translation keys

Dragon-Seeker committed:
Add more docs and include some annotations for various classes in config

Noaaan commented:
Despite missing the last task (Make custom font for better looking clipboard), do you think this is ready for review? Might make the overall review process faster in case changes are needed

Dragon-Seeker commented:
> Despite missing the last task (Make custom font for better looking clipboard), do you think this is ready for review? Might make the overall review process faster in case changes are needed

Overall, it's ready for initial reviews by people, as the reason for it being a draft is that when I opened the PR, I still needed to add some changes. I also planned to wait for glisco to do an initial review of the changes, as I felt they may have some issues with the changes.

Dragon-Seeker committed:
Merge branch '1.21.9' into 1.21.6-ConfigAdjustments

Dragon-Seeker committed:
Update for Config PR for 1.21.10

Dragon-Seeker committed:
Update EndecArgumentSerializer for yarn and use constant argument serializer

Dragon-Seeker committed:
Sweeping fixes with updating docs, adjusting code and overall cleanup

- Adjust Constraints to have a type to indicate to users of the API what each predicate is used for
- Switch to only using endec for serialization and abstracting serialization of the given config wrapper to a ConfigSerializer allowing for other formats than just Jankson
- Update ModMenu annoation to use the modid of the config wrapper annotation
- Cleanup predicate handling for OptionComponents and use more statically defined methods
- Refactor location of code and remove commented out code combined with making code prettyer
